### DISCUSSION TITLE
The Android app stops receiving packets from the gateway

### DISCUSSION THREAD

**Discussion by gongoscho** [strongswan/strongswan#2234] [General]
I've been having some issues with the Android application.

It stops receiving packets from the gateway, but after the roadwarrior went offline (after trying for 5 retransmits), the sessions from the roadwarrior are seen on the firewall (from 11:01:25 12[IKE] sending keep alive to WAN_IP_STATIC[4500] in the rw_android.log).

The VPN is showing as still connected but not working. I've attached the rw_android.log
The app isn't set up to save energy.

My system:
Gateway:
Firewall OPNsense 24.1.6-amd64
FreeBSD 13.2-RELEASE-p11
OpenSSL 3.0.13
strongSwan 5.9.13

Roadwarior:
Xiaomi Poco F1 LineageOS 20
Android 13
strongSwan VPN Client 2.4.2 (the latest Version from F-Droid Store)
strongSwan 5.9.11

The configurations
swanctl.conf:
```
# This file is automatically generated. Do not edit
connections {
    6c590eaf-f14b-4029-b98b-42c1086f96fd {
        proposals = aes256gcm16-sha512-x25519
        unique = no
        aggressive = no
        version = 2
        mobike = yes
        encap = no
        dpd_delay = 30
        pools = pool-roadwarrior-ipv4
        send_certreq = no
        local-90c22aef-d2bd-42fd-82fc-2ecff73752fd {
            round = 0
            auth = pubkey
            certs = 66034635e0793.crt
        }
        remote-89101440-1882-4c18-94be-0c3021ceb290 {
            round = 0
            auth = eap-mschapv2
            eap_id = %any
        }
        children {
            d68f303c-d4cd-436d-ac2c-4d5dbb782f40 {
                esp_proposals = aes256gcm16-x25519
                sha256_96 = no
                start_action = none
                close_action = none
                dpd_action = clear
                mode = tunnel
                policies = yes
                local_ts = 0.0.0.0/0
                rekey_time = 3600
                updown = /usr/local/opnsense/scripts/ipsec/updown_event.py --connection_child d68f303c-d4cd-436d-ac2c-4d5dbb782f40
            }
        }
    }
}
pools {
    pool-roadwarrior-ipv4 {
        addrs = 10.0.1.0/30
        dns = 10.0.0.1
    }
}
secrets {
    eap-04e3f318-ddcd-4336-a662-3e2341bd913f {
        id-0 = ***@***.**
        secret = ***
    }
    eap-cfed74c7-7b04-4e65-a77c-67406002b6ae {
        id-0 = ***@***.**
        secret = ***
    }
}
# Include config snippets
include conf.d/*.conf
``` 

Logs
[ipsec.log](https://github.com/strongswan/strongswan/files/15210799/ipsec.log)

[rw_android.log](https://github.com/strongswan/strongswan/files/15210826/rw_android.log)

I've also tested with forced UDP encapsulation and without dead peer detection, but unfortunately I got the same result.

Could you help me out?

grEEtZ

**Comment by tobiasbrunner** [comment]
Consider changing the [retransmission settings](https://docs.strongswan.org/docs/5.9/config/retransmission.html) and possibly other timed events (e.g. DPDs, rekeyings) on the server so it doesn't remove the SAs while clients are offline.  The client might not notice that as it does not send DPDs actively, so the server should be configured to only remove completely abandoned SAs so clients can be offline for a while.

Also, check power settings etc. on the client so the Android system does not disable the network connectivity for the app or put the app to sleep prematurely.

**Comment by gongoscho** [comment]
Hello,

i finally solved the issue. I had to change the NAT-T keepalive interval in the Android app from 45 seconds to 23 seconds and I had to set the same algorithms / "ESP proposals" on the client as on the gateway because the rekeying did not work. Error in log: "NO_PROPOSAL_CHOSEN".

With this setting, the client will reconnect if the VPN gateway is restarted or stopped/started, or the client is offline or its IP has changed (MOBIKE).
 
Do I need to set my DPD Delay to at least twice the value of the NAT Keepalive or the same value to ensure a secure buffer?

In this case, a DPD delay of at least 23 seconds would be correct?
With a DPD delay of 23 seconds, I do not see any NAT-T keepalives from the client.


### **My understanding**
**DPD (Dead Peer Detection)** in IKEv2 is a mechanism used to determine if the other endpoint of the VPN connection is still active. If there is no response to DPD requests, the connection is considered dead and can be reestablished. This is important to ensure that the connection is quickly reestablished if it is lost.

**NAT Keepalive** is used to keep the entries in the router's NAT table up to date. Since NAT entries can expire over time, the Roadwarrior will periodically send small packets to keep the connection open. This prevents the connection from being lost due to NAT timeouts.



grEEtZ

**Reply by tobiasbrunner** [reply]
> I had to set the same algorithms / "ESP proposals" on the client as on the gateway because the rekeying did not work. Error in log: "NO_PROPOSAL_CHOSEN".

What exactly was the mismatch? The client includes proposals with and without common DH groups, so there shouldn't be a mismatch (unless a rare DH group was chosen by the server).



> With this setting, the client will reconnect if the VPN gateway is restarted or stopped/started, or the client is offline or its IP has changed (MOBIKE).

Not sure how the NAT-T keepalive helps here exactly. Maybe it improves the chances that the delete from the server reaches the client. But it should not have an effect on MOBIKE as that will trigger outbound messages anyway.


> Do I need to set my DPD Delay to at least twice the value of the NAT Keepalive or the same value to ensure a secure buffer?

I usually recommend to set it as high as possible to account for the client being offline/unreachable for longer intervals (the DPDs would then just be used to remove actually abandoned SAs, but that might already happen anyway if the client reconnects and sends an INITIAL_CONTACT notify). If the server destroys the SA relatively quickly, the client won't be aware of it once it is reachable again. As I wrote above, the same goes for the rekeying intervals as the those exchanges will act as DPDs as well.


> With a DPD delay of 23 seconds, I do not see any NAT-T keepalives from the client.

NAT-T keepalives are sent if no other messages (IKE or ESP) were sent within the configured interval. As each DPD is an exchange there is a message sent and no keepalive is necessary.


> If there is no response to DPD requests, the connection is considered dead and can be reestablished.

Not from the server. Only the client could reestablish it (but it won't be aware of it if the server unilaterally destroys the SA due to a DPD timeout).

**Reply by gongoscho** [reply]
> What exactly was the mismatch? The client includes proposals with and without common DH groups, so there shouldn't be a mismatch (unless a rare DH group was chosen by the server).

I am using only one DH32 group on the VPN gateway. Changed from Curve25519 to Curve448.
If i only remove the following in the Android app:

- aes256gcm16-sha512-curve448
- aes256gcm16-curve448


I get the following error in the log and the client is also offline according to the log. 

> creating rekey job for CHILD_SA ESP/0xc1ffe17e/WAN_VPNGW-IP
> received NO_PROPOSAL_CHOSEN notify, no CHILD_SA built
> failed to establish CHILD_SA, keeping IKE_SA
> CHILD_SA rekeying failed, trying again in 9 seconds
> received DELETE for IKE_SA
> deleting IKE_SA
> IKE_SA deleted
> lease went offline
> parsed IKE_SA_INIT request 0 
> WAN_VPNCLIENT-IP is initiating an IKE_SA
> selected proposal: IKE:AES_GCM_16_256/PRF_HMAC_SHA2_512/CURVE_448
>  DH group ECP_256 unacceptable, requesting CURVE_448
> ...

I attached: [ipsec.log](https://github.com/user-attachments/files/17412445/ipsec.log)


> I usually recommend to set it as high as possible to account for the client being offline/unreachable for longer intervals (the DPDs would then just be used to remove actually abandoned SAs, but that might already happen anyway if the client reconnects and sends an INITIAL_CONTACT notify). 

In this case, a DPD delay of at least 46 seconds would be correct, when i set the NAT-T keepalive to 23?
What value would you recommend?

> NAT-T keepalives are sent if no other messages (IKE or ESP) were sent within the configured interval. As each DPD is an exchange there is a message sent and no keepalive is necessary.

> Not from the server. Only the client could reestablish it (but it won't be aware of it if the server unilaterally destroys the SA due to a DPD timeout).

Thanks for the education

**Reply by tobiasbrunner** [reply]
> I am using only one DH32 group on the VPN gateway. Changed from Curve25519 to Curve448.

I see. Yeah, `curve448` is currently not part of the proposal sent by the Android client (see [here](https://github.com/strongswan/strongswan/blob/faf40b8d746c4daed148c573443d7fb10fbea7c7/src/frontends/android/app/src/main/jni/libandroidbridge/backend/android_service.c#L877)). It wasn't supported for a long time by the app, now that we use OpenSSL 3 as crypto library it is. I'll consider adding it to the proposal in a future update.

> In this case, a DPD delay of at least 46 seconds would be correct, when i set the NAT-T keepalive to 23?

No, I'm talking about DPD delays of hours, not seconds. Really depends on the usage pattern though (how long you expect the client to stay connected and in particular how long it might be unreachable due to roaming between networks or because Android puts itself to sleep).

**Reply by gongoscho** [reply]
> I see. Yeah, `curve448` is currently not part of the proposal sent by the Android client (see [here](https://github.com/strongswan/strongswan/blob/faf40b8d746c4daed148c573443d7fb10fbea7c7/src/frontends/android/app/src/main/jni/libandroidbridge/backend/android_service.c#L877)). It wasn't supported for a long time by the app, now that we use OpenSSL 3 as crypto library it is. I'll consider adding it to the proposal in a future update.

Should I create an enhancement: "Add Curve448 to the function: Create ESP proposals with and without DH groups - in the Android app" and refer to this discussion?

> No, I'm talking about DPD delays of hours, not seconds. Really depends on the usage pattern though (how long you expect the client to stay connected and in particular how long it might be unreachable due to roaming between networks or because Android puts itself to sleep).


> Not sure how the NAT-T keepalive helps here exactly. Maybe it improves the chances that the delete from the server reaches the client. But it should not have an effect on MOBIKE as that will trigger outbound messages anyway.

I can't get the DPD and NAT - Keepalive Interval out of my head and I want to understand the problem.

If I leave the NAT keepalive interval in the Android app at the default value of: 45 seconds and set DPD e.g. to 9 seconds (idea: 9 x 5 default retransmissions = 45, if there is no contact by then, the client is offline), then there is no problem, because every 9 seconds, if no data is sent from the client, there is always a successful DPD - INFORMATIONAL response from the Android app.

But sending a packet every 9 seconds seems to me to be too much traffic. Too bad about the resources
Unfortunately, I haven't had time to test for longer than 9 seconds with NAT keepalive of 45 seconds.

However, if I set the DPD interval to 300, the problem remains as in my first post.

**Reply by tobiasbrunner** [reply]
> Should I create an enhancement: "Add Curve448 to the function: Create ESP proposals with and without DH groups - in the Android app" and refer to this discussion?

No, that's fine, I've added it to my internal TODO list.



> However, if I set the DPD interval to 300, the problem remains as in my first post.

I understood it works if you reduce the NAT keepalive interval on the client, no? Since expiring NAT mappings is more of a client issue, this seems like a reasonable thing to do (note that strongSwan uses 20 seconds by default, the app [uses a higher default](https://github.com/strongswan/strongswan/commit/73a6bec3fcd9154123cd2f726b52eaeb677acb12) in an attempt to reduce the impact on battery life), or change the NAT settings in the affected router if you can.

**Reply by gongoscho** [reply]
>  No, that's fine, I've added it to my internal TODO list.
> 

Ok, thank you! :1st_place_medal: 

> I understood it works if you reduce the NAT keepalive interval on the client, no? 
>
Yes :+1: 

With dpd of 0, I found the correct nat keepalive setting that the client always has a stable connection.

I can see the connection on the firewall / VPN gateway, and the firewall is also set to try to avoid dropping legitimate idle connections ([firewall optimization](https://docs.opnsense.org/manual/firewall_settings.html#firewall-optimization): conservative). I also tested other firewall optimization modes, but there was no difference.

I also test to set dpd to default of 0 (disable), and with higher nat-keep alives there is the same situation that the connection looks online on the android client and VPN gateway, but does not work. And you see the keep alives in the app/log on the android client.

And the keep alives / active connection I see on the firewall. Strange.

> change the NAT settings in the affected router if you can.



I can't change the NAT settings on the client in the client network, as I'm only testing on cellular / 4G.

I don't think I need to change "[charon.retransmit_tries](https://docs.strongswan.org/docs/5.9/config/retransmission.html)" with a value of 5, because it shouldn't come to that since the client is online and sending NAT keep alives.

Does the setting: "[<conn>.keyingtries](https://docs.strongswan.org/docs/5.9/swanctl/swanctlConf.html#_connections)" in "swanctl.conf" on the VPN gateway actually affect the connection on the Android client?

It defaults to 1, I have often seen that you should set it to 0. 

A value of 0 initiates a new sequence until the connection is established or fails with a persistent error. But is this only important for site-to-site connections?

**Reply by tobiasbrunner** [reply]
> I also tested other firewall optimization modes, but there was no difference.

The server side's firewall configuration won't make a difference. It's a problem with NAT mappings timing out relatively quickly on a router/firewall near the client.



> Does the setting: "[.keyingtries](https://docs.strongswan.org/docs/5.9/swanctl/swanctlConf.html#_connections)" in "swanctl.conf" on the VPN gateway actually affect the connection on the Android client?

No, this setting is only used as initiator. The gateway won't be able to reestablish the connection once the client isn't reachable anymore (usually the asymmetric authentication/configuration settings prevent it anyway), so this won't have any effect.

**Reply by gongoscho** [reply]
I have discovered something shocking.

Since I cannot change the IPSec NAT-T Port: 4500 UDP via the GUI in the firewall: OPNsense, a NAT is created on WAN: HTTP 80 UDP to NAT-T 4500 UDP.

I have set the NAT keepalive back to default 45 via port 80 and no more problems...

Same with Port 443 UDP(QUIC)

**Reply by gongoscho** [reply]
> > I also tested other firewall optimization modes, but there was no difference.
> 
> The server side's firewall configuration won't make a difference. It's a problem with NAT mappings timing out relatively quickly on a router/firewall near the client.
> 
> > Does the setting: "[.keyingtries](https://docs.strongswan.org/docs/5.9/swanctl/swanctlConf.html#_connections)" in "swanctl.conf" on the VPN gateway actually affect the connection on the Android client?
> 
> No, this setting is only used as initiator. The gateway won't be able to reestablish the connection once the client isn't reachable anymore (usually the asymmetric authentication/configuration settings prevent it anyway), so this won't have any effect.

Hello,

I'm desperate. I set it back to UDP 4500 (default) and with NAT-T keepalive interval 23/20, I get the same problem after a few hours.

I only noticed it overnight.

I tested it again without DPD except for NAT-T keepalive interval 10.
It only got worse. There is no setting lower than 10 on the Android client.

The configuration has remained almost the same:

I have now set the rekey time to 1440 seconds (default 4 hours) and phase 2 to 360 seconds (default 3600).
On Android Client: proposals = aes256gcm16-sha512-x448 & esp_proposals = aes256gcm16-x448

The rekeying can't be the problem because it often runs without any issues.

Updated Systems:

Android 15 | LineageOS for microG 22.2 
OPNsense: OPNsense 25.7.9
strongSwan 6.0.3
Android app: 2.6.2 (F-Droid)

The Android app is set so that it doesn't save energy.


Could you give me some tips on what I could set to identify the problem?

Thank you

**Reply by tobiasbrunner** [reply]
> Could you give me some tips on what I could set to identify the problem?

Not without logs from both ends that show the complete session.

**Reply by gongoscho** [reply]
Hello,

here are the logs:
[charon_android.log](https://github.com/user-attachments/files/23980384/charon_android.log)
[ipsec.log](https://github.com/user-attachments/files/23980385/ipsec.log)


Start at 05:40 on screen off on Android.
Test the internet with messages over signal from another Android.

First message sent at 05:41:10 and received at 05:43:10.

Second message sent at 05:46:30 and received at 05:46:30.

Third message sent at 05:53:34 and received at 05:53:34.


Fourth message sent at 06:06:01 and received at 06:06:01.

Fifth message sent at 06:26:33, reception no longer possible.

Then turned on the screen on the Android and tested the VPN connection with the browser. Connection failed. However, the VPN is always connected.

Log pulled from Android. strongSwan switched off on the FW and log pulled.

IPs, username, and certificate anonymized in the logs.

In the FW logs, I noticed that I still had an old certificate stored where the CN no longer matches the current IP.

I have also changed my ISP in the meantime.

It's interesting that a connection is still possible even though the CN in the certificate does not match the IP on the Android device.

**Reply by tobiasbrunner** [reply]
> Fifth message sent at 06:26:33, reception no longer possible.
> 
> Then turned on the screen on the Android and tested the VPN connection with the browser. Connection failed. However, the VPN is always connected.

The client won't notice if the server closes the connection. Which it does at 06:26:27 because it can't reach the client for an IKE SA rekeying. So it will show as connected but any packets it sends won't be processed by the server as it already removed the SAs.

The last log message on the client is at 06:19:11. It apparently doesn't send any keepalives afterwards. And that causes the server to fail the rekeying it starts at 06:23:42. Why it would just stop I don't know. Could be some hardcore Android energy saving mechanism (no idea if LineageOS does anything special/different in that regard or if it even depends on the hardware). Also, it should send keepalives again if you wake the device and open the app (maybe you didn't wait long enough for the delayed timeout to elapse). If you prevent the server from initiating rekeyings or DPDs (or only after several hours) you should at least be able to use the VPN after waking up the device.

> It's interesting that a connection is still possible even though the CN in the certificate does not match the IP on the Android device.

The CN doesn't matter, only SANs are relevant.

**Reply by gongoscho** [reply]
Yes, I hope it's not a LineageOS or hardware issue. Thanks for the initial analysis!

I will improve my tests.

Should I change the time values from the rekeying in the phases so that it is possible to determine exactly when Android is in "deep sleep" and no longer responding?

If I notice that the client is unavailable, I will try to test the connection with the browser for longer.

Does it make sense to test the connection until the next rekeying (in my case 360 seconds)?

**Reply by tobiasbrunner** [reply]
> Should I change the time values from the rekeying in the phases so that it is possible to determine exactly when Android is in "deep sleep" and no longer responding?

I don't think that would be all that useful. Maybe the system log (logcat) has some information about power stuff.

> Does it make sense to test the connection until the next rekeying (in my case 360 seconds)?

If the server already removed the SAs, there won't be a rekeying.
